### ISSUE TITLE
how to run on a new dataset?

### ISSUE THREAD

**Issue by YonghanLU** [mathpluscode/ImgX-DiffSeg#18]
As the title, I am not familiar with jap and tf. Looking forward to your advice. By the way, my experiments will be conducted on Linux and a single GPU using PyCharm.

**Comment by mathpluscode** [comment]
Hi @YonghanLU, thanks for your question.

I understand that it may not be obvious as we are using [tfds](https://www.tensorflow.org/datasets/api_docs/python/tfds). Let me try to briefly explain how this would work, and the four existing datasets could potentially also be helpful for illustrations. Meanwhile, for JAX/Flax, I would recommend checking [Flax examples](https://flax.readthedocs.io/en/latest/examples/index.html) which are great resources.

The main steps of adding a custom dataset with tfds for segmentation would be:
1. create a new folder in `imgx_datasets/` https://github.com/mathpluscode/ImgX-DiffSeg/tree/main/imgx_datasets such as `demo`.
2. under `imgx_datasets/custom_dataset` create `demo_dataset_builder.py`, inside which we need to create a new class `Builder`, similar to [male_pelvic_mr_dataset_builder.py](https://github.com/mathpluscode/ImgX-DiffSeg/blob/main/imgx_datasets/male_pelvic_mr/male_pelvic_mr_dataset_builder.py), you can also read the TensorFlow documentation https://www.tensorflow.org/datasets/add_dataset. The implementation does not require much TensorFlow knowledge. The main requirement is that `_generate_examples` should `yield uid, {UID: uid, IMAGE: image[..., None], LABEL: label}` there the `IMAGE` should have (spatial_dimensions, input_channel).

Technically it is not a hard requirement as the JAX/FLAX framework is compatible with any data loaders. Together with #15, I understand that some toy examples/usages could potentially be more helpful. I will try my best to provide some examples in the coming weeks.

**Comment by YonghanLU** [comment]
@mathpluscode Thanks. Another question also needs your help. In the reverse process, why only need 5 steps? I am not clear from your two papers and code. Is it randomly sampled 5 values from [T-1, 0] ? Can you give me a detailed pytorch-version source about this sampling method?

**Comment by mathpluscode** [comment]
> @mathpluscode Thanks. Another question also needs your help. In the reverse process, why only need 5 steps? I am not clear from your two papers and code. Is it randomly sampled 5 values from [T-1, 0] ? Can you give me a detailed pytorch-version source about this sampling method?

Dear @YonghanLU, assuming your "reverse process" is the denoising process (right to left in the figure below)
<img width="385" alt="image" src="https://github.com/mathpluscode/ImgX-DiffSeg/assets/4066929/5401068f-0696-45e1-8581-07b5ebf4bf9a">

The corresponding code is at https://github.com/mathpluscode/ImgX-DiffSeg/blob/main/imgx/diffusion/gaussian/variance_schedule.py#L101. Although these codes are in JAX, we are using `jax.numpy` which has exactly the same signature as numpy. The relevant literature is https://arxiv.org/abs/2102.09672

<img width="430" alt="image" src="https://github.com/mathpluscode/ImgX-DiffSeg/assets/4066929/8886c90d-a8cb-4c95-906b-9ddf42ca2420">

In a nutshell, given 1001 steps, we sample 5 steps evenly, 1, 251, 501, 751, 1001, then uses `alphas_cumprod` or $\bar{\alpha}_{t}$ for $t = 1,251,501,751,1001$ to redefine the betas

<img width="322" alt="image" src="https://github.com/mathpluscode/ImgX-DiffSeg/assets/4066929/5eca58e9-7767-4ba6-b1bf-2bedc92d2d4d">

⚠️ Note, it is different from the noising schedule here https://github.com/openai/guided-diffusion/blob/main/guided_diffusion/gaussian_diffusion.py#L18. However, as mentioned this is a choice and this may be application dependant.

Such a design can be considered as a hyper-parameter. You may be interested in the following literature which discusses more noising schedules
- https://arxiv.org/abs/2301.10972
- https://arxiv.org/abs/2305.08891

Please let me know if this is still unclear for you :)

**Comment by YonghanLU** [comment]
Thanks for your detailed explanation! Btw, the training process and testing process both use the resampling strategy or only the testing process ?

**Comment by mathpluscode** [comment]
> Thanks for your detailed explanation! Btw, the training process and testing process both use the resampling strategy or only the testing process ?

Hi @YonghanLU, the resampling is certainly used for testing/inference for performance consideration. For training, it was used in our workshop paper https://arxiv.org/abs/2303.06040 as it showed improvements (Table 3.c), however in our latest work https://www.melba-journal.org/papers/2023:016.html, we only observed the improvements in the muscle ultrasound data set but not in the abdominal CT data set (Table 10 in appendix F.3). We thus did not apply the resampling for training in this work. This work only focuses on the recycling strategy itself.

However, this is indeed an interesting observation and this potentially may be linked to the noising schedule, noising method, application (network, data, etc.).
